### PR TITLE
PERF: Adding weights caching derivative calculation to iktMeanSquaresImageToImageMetric

### PR DESCRIPTION
**Pull request by ljm898**
Weights caching is used for derivative calculation to reduce registration time with BSplineTransforms. Currently only implemented for MeanSquaresImageToImageMetric, other metrics to be added at a later date.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@ljm898 You may have missed some modifications:

[Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:169](https://github.com/InsightSoftwareConsortium/ITK/blob/759738c4b02e21ba0e4fd94183a96c5544f799fc/Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx#L169):34: error: use of undeclared identifier 'm_NumBSplineWeights'


https://open.cdash.org/viewBuildError.php?buildid=10043237

The code is not compiling for this PR.

**Comment by ljm898** [comment]
Thanks in advance for your patience! Not sure why it worked on my pc when I tested it. From context and the error, it looks like it was missing a this-> at those lines. Hopefully that fixes it.

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx [hunk 1]
```diff
@@ -164,16 +164,30 @@ MeanSquaresImageToImageMetric<TFixedImage, TMovingImage>::GetValueAndDerivativeT
     transform = this->m_Transform;
   }
 
-  // Jacobian should be evaluated at the unmapped (fixed image) point.
-  transform->ComputeJacobianWithRespectToParameters(fixedImagePoint, threadS.m_Jacobian);
-  for (unsigned int par = 0; par < this->m_NumberOfParameters; ++par)
+  if (this->m_TransformIsBSpline && this->m_UseCachingOfBSplineWeights)
```
Comment (hjmjohnson) [inline Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:167]: this->m_TransformIsBSpline is deprecated.  

```
 \deprecated `m_TransformIsBSpline` is intended to be removed, in the future. Please use `m_BSplineTransform`
```

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx [hunk 2]
```diff
@@ -164,16 +164,30 @@ MeanSquaresImageToImageMetric<TFixedImage, TMovingImage>::GetValueAndDerivativeT
...
+  { // using pre-computed weights and indexes to calculate only non zero elements of the derivative
+    for (unsigned int k = 0; k < this->m_NumBSplineWeights; ++k)
+    {
+      for (unsigned int dim = 0; dim < MovingImageDimension; ++dim)
+      {
+        int par = this->m_BSplineTransformIndicesArray[fixedImageSample][k] + this->m_BSplineParametersOffset[dim];
+        threadS.m_MSEDerivative[par] += 2.0 * diff * this->m_BSplineTransformWeightsArray[fixedImageSample][k] * movingImageGradientValue[dim];
```
Comment (hjmjohnson) [inline Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:174]: I changed line 174 to 
threadS.m_MSEDerivative[par] = 99;

and no tests failed. 


We need to have a tests that executes this code before it can be added to the toolkit.  Currently the code is never run by any tests.

**Review by hjmjohnson — CHANGES_REQUESTED** [review]
Please make this a single commit on top of the current master branch.  Having two commits for  1 change makes it very hard to evaluate the change.


Recommended code:
```c++
  if (this->m_BSplineTransform && this->m_UseCachingOfBSplineWeights)
  {
    // using pre-computed weights and indexes to calculate only non zero elements of the derivative
    for (unsigned int w = 0; w < this->m_NumBSplineWeights; ++w)
    {
      const auto precomputedIndex = this->m_BSplineTransformIndicesArray[fixedImageSample][w];
      const auto precomputedWeight = this->m_BSplineTransformWeightsArray[fixedImageSample][w];
      for (unsigned int dim = 0; dim < MovingImageDimension; ++dim)
      {
        const int par = precomputedIndex + this->m_BSplineParametersOffset[dim];
        //threadS.m_MSEDerivative[par] += 2.0 * diff * precomputedWeight * movingImageGradientValue[dim];
        //Need to verify that a test fails with wrong answer below.
        threadS.m_MSEDerivative[par] += 99.0 * diff * precomputedWeight * movingImageGradientValue[dim];
      }
    }
  }
  else
  {
...
```

[ ] - Need to ensure that a test fails with a bad computation in the caching component.

**Comment by ljm898** [comment]
Thank you for your review, I very much appreciate your time. I will get to this in the coming week.

**Comment by hjmjohnson** [comment]
@ljm898 thank you!  

Your efforts are greatly appreciated.

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx (hunk 2 above)
Comment (ljm898) [reply Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:174]: Test has been added to itkMeanSquaresImageMetricTest.cxx that checks consistency between derivative computed omitting versus using weights caching. Tested on my end, can confirm it fails using 99.0 * ... and passes using 2.0 * ...

Comment (hjmjohnson) [reply Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:174]: @ljm898 We are looking forward to seeing these changes :).

**Review by hjmjohnson — CHANGES_REQUESTED** [review]

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx (hunk 2 above)
Comment (ljm898) [reply Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:174]: Looks like the code didn't push properly. Working on it now.

**Comment by ljm898** [comment]
Something has gone wrong on my end...

**Comment by ljm898** [comment]
Trying to reopen this pull request

**Comment by ljm898** [comment]
Something has gone wrong on my end

**Comment by ljm898** [comment]
Trying to reopen this

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx [hunk 3]
```diff
@@ -81,7 +81,7 @@ MeanSquaresImageToImageMetric<TFixedImage, TMovingImage>::GetValueThreadProcessS
   const MovingImagePointType & itkNotUsed(mappedPoint),
   double                       movingImageValue) const
 {
-  const double diff = movingImageValue - this->m_FixedImageSamples[fixedImageSample].value;
+  double diff = movingImageValue - this->m_FixedImageSamples[fixedImageSample].value;
```
[the remark this answers is not in the document]
Comment (dzenanz) [reply Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:84]: I think this code was from before the commit which added `const` qualifier wherever feasible. Maybe a rebase on top of current master would take care of it?

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 4]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
...
+  {
+    std::cout << "Failed to initialise metric to test BSplineTransform derivative computation omitting weights caching."
+              << std::endl;
+    std::cout << "Reason " << e.GetDescription() << std::endl;
+
+    return EXIT_FAILURE;
+  }
+
+  // Initialise metric using caching
+  auto metricWithCaching = MetricType::New();
```
Comment (dzenanz) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:354]: Why not reuse variable `metricNoCaching, e.g.:
```cpp
MetricType::DerivativeType derivativeNoCaching;
metricNoCaching->GetDerivative(bSplineParameters, derivativeNoCaching);

metricNoCaching->SetUseCachingOfBSplineWeights(true);
MetricType::DerivativeType derivativeWithCaching;
metricNoCaching->GetDerivative(bSplineParameters, derivativeWithCaching);
```
And rename it in that case to e.g. `metric`. Having two variables would make sense if we evaluated derivates many times.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 5]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
...
+  BSplineTransformType::MeshSizeType meshSize;
+  meshSize.Fill(nodesPerDimension - splineOrder);
+  initializer->SetTransform(bSplineTransform);
+  initializer->SetImage(movingImage);
+  initializer->SetTransformDomainMeshSize(meshSize);
+  initializer->InitializeTransform();
+
+  // Set bSplineTransform parameters
+  ParametersType bSplineParameters(bSplineTransform->GetNumberOfParameters());
+  bSplineParameters.Fill(0.0);
```
Comment (dzenanz) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:329]: I am not sure whether all-zero parameters will reveal differences.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 4 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:354]: I may have been overly cautious. I wanted to be certain that the derivatives being set by reference were different, not just pointing to the same buffer. Something along the lines of trying to be correct and redundant rather than efficient but wrong.

File: Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ljm898) [reply Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx:84]: Can confirm this pr was initiated before that commit, happy to rebase.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 5 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:329]: I'm not sure I understand what you're asking here. Do you mean that checking only the consistency between cached vs non-cached derivatives is insufficient to catch differences? Or that initialising the transform with all zero parameters may return zero derivatives?

As far as the suggested test goes, the moving image offset is changed from (0,0) to (1,1) at the beginning of the test, while the fixed image keeps its offset as (0,0). If my understanding is correct, this should be sufficient to ensure derivatives are non-zero given that the interpolation weights are being used during derivative calculation. Either way, I'm happy to initialise the parameters with non-zero values?

Comment (dzenanz) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:329]: > initialising the transform with all zero parameters may return zero derivatives

This is what I am afraid of.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 6]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
...
+    metricNoCaching->Initialize();
+  }
+  catch (const itk::ExceptionObject & e)
+  {
+    std::cout << "Failed to initialise metric to test BSplineTransform derivative computation omitting weights caching."
+              << std::endl;
+    std::cout << "Reason " << e.GetDescription() << std::endl;
+
+    return EXIT_FAILURE;
+  }
```
Comment (jhlegarreta) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:351]: Use `ITK_TRY_EXPECT_NO_EXCEPTION( metricNoCaching->Initialize());` instead of https://github.com/InsightSoftwareConsortium/ITK/pull/4982/files#diff-c3d685418473012ef3d09730cf43126a837aba1b9e8cc7e63cc1cd09616c0239R342-R351

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 7]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
...
+    metricWithCaching->Initialize();
+  }
+  catch (const itk::ExceptionObject & e)
+  {
+    std::cout << "Failed to initialise metric to test BSplineTransform derivative computation using weights caching."
+              << std::endl;
+    std::cout << "Reason " << e.GetDescription() << std::endl;
+
+    return EXIT_FAILURE;
+  }
```
Comment (jhlegarreta) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:372]: Use `ITK_TRY_EXPECT_NO_EXCEPTION(metricWithCaching->Initialize());` instead of https://github.com/InsightSoftwareConsortium/ITK/pull/4982/files#diff-c3d685418473012ef3d09730cf43126a837aba1b9e8cc7e63cc1cd09616c0239R361-R372

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 8]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
   //-------------------------------------------------------
   metric->Print(std::cout);
 
+
+   //-------------------------------------------------------
```
Comment (jhlegarreta) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:302]: I know these empty comment lines are being used already in the test, but please remove them from the new code.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 9]
```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
...
+  // Check consistency between BSplineTransform
+  // derivatives computed with vs omitting weights caching.
```
Comment (jhlegarreta) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:304]: The line length may allow for the commend to dwell on a single line.

**Review by jhlegarreta — COMMENTED** [review]
Some style comments, but mainly the adoption of `ITK_TRY_EXPECT_NO_EXCEPTION`.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 4 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:354]: Not sure whether I'll finish revising code changes today - I'm planning to randomly initialise each parameter with MersenneTwister. Will that be ok?

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 5 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:329]: Not sure whether I'll finish revising code changes today - I'm planning to randomly initialise each parameter with MersenneTwister. Will that be ok?

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 4 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:354]: Code being revised to use single metric object to test caching vs no caching.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 6 above)
Comment (ljm898) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:351]: Thank you for this suggestion! The metricNoCaching and MetricWithCaching are being combined into a single metric following another suggestion - ITK_TRY_EXCEPT_NO_EXCEPTION is used there.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx [hunk 10]
```diff
@@ -39,7 +43,7 @@ itkMeanSquaresImageMetricTest(int, char *[])
 
   // Save the format stream variables for std::cout
   // They will be restored when coutState goes out of scope
-  const itk::StdStreamStateSave coutState(std::cout);
+  itk::StdStreamStateSave coutState(std::cout);
```
Comment (dzenanz) [inline Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx:46]: This and a few other variables, unrelated to changes, lost their const-ness.

**Review by dzenanz — COMMENTED** [review]
Mostly looks good to me.

File: Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx (hunk 6 above)
Comment (dzenanz) [reply Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx]: I think you need to `#include "itkTestingMacros.h"`

**Comment by ljm898** [comment]
Had a look into the failed/failing ITK.macOS check. itkNiftiReadWriteDirectionTest fails with code 255. I'd have expected the failed test to be independent of the suggested code changes. Any suggestions for fixing this compilation error?

**Comment by N-Dekker** [comment]
> Had a look into the failed/failing ITK.macOS check. itkNiftiReadWriteDirectionTest fails with code 255. I'd have expected the failed test to be independent of the suggested code changes. Any suggestions for fixing this compilation error?

You may rerun the CI for this machine, by adding a comment, saying something like `/azp run <machine-name>`. Well... I'll just give it a try  😃

**Review by dzenanz — APPROVED** [review]
LGTM

**Comment by N-Dekker** [comment]
Just for my understanding, it looks like the `UseCachingOfBSplineWeights` property has been around for decades already (looking at git log). Did the original developers somehow forget to use it?

In which scenario's would you suggest users to switch it on?

**Comment by ljm898** [comment]
> Just for my understanding, it looks like the `UseCachingOfBSplineWeights` property has been around for decades already (looking at git log). Did the original developers somehow forget to use it?
> 
> In which scenario's would you suggest users to switch it on?

Current versions use caching by default. AFAIK, caching was only used to compute deformation instead of the transform point method. Not sure why it wasn't originally used for derivatives - it could be the case that additional support has since been added to ITK that made it practical? I'm just stoked it works for derivatives.

**Review by hjmjohnson — APPROVED** [review]
